gce_workload_certs_refresh/main.go

```diff
@@ -0,0 +1,290 @@
+//  Copyright 2022 Google LLC
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+// GoogleAuthorizedKeys obtains SSH keys from metadata.
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"time"
+
+	"github.com/GoogleCloudPlatform/guest-logging-go/logger"
+)
+
+const (
+	contentDirPrefix  = "/run/secrets/workload-spiffe-contents"
+	tempSymlinkPrefix = "/run/secrets/workload-spiffe-symlink"
+	symlink           = "/run/secrets/workload-spiffe-credentials"
+)
+
+var (
+	programName    = "gce_workload_certs_refresh"
+	version        string
+	metadataURL    = "http://169.254.169.254/computeMetadata/v1/"
+	defaultTimeout = 2 * time.Second
+)
+
+func logFormat(e logger.LogEntry) string {
+	now := time.Now().Format("2006/01/02 15:04:05")
+	return fmt.Sprintf("%s: %s", now, e.Message)
+}
+
+func getMetadata(key string) ([]byte, error) {
+	client := &http.Client{
+		Timeout: defaultTimeout,
+	}
+
+	url := metadataURL + key
+	req, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Add("Metadata-Flavor", "Google")
+
+	var res *http.Response
+	// Retry forever, increase sleep between retries (up to 5 times) in order
+	// to wait for slow network initialization.
+	var rt time.Duration
+	for i := 1; ; i++ {
+		res, err = client.Do(req)
+		if err == nil {
+			break
+		}
+		if i < 6 {
```

Comment:
why 6?

Reply:
it's semi arbitrary. the idea is that for the first 6 rounds of waiting, we will continually increase the sleep time. this is a progressive backoff. but at that point it will be 15 seconds, which we decided is the maximum backoff time.

the real reason is because i copied this code from one of the other binaries :) probably the first half of this file is common stuff. ideally maybe we would have a kind of envelope library that makes it all common, but it hasn't been pressing up til now.

Reply:
you're implementing a timer logic inside a service that is meant to be executed in a timer context, make it linear trying n times and waiting m seconds between each attempt and let systemd timer logic do their work for you, end the execution and try it later - keep it simple.

Reply:
done